File: sitemap_parser.py
"""
Sitemap Parser - nalezení a parsování sitemapy z webu
"""
import requests
import xml.etree.ElementTree as ET
from urllib.parse import urljoin, urlparse
from typing import List, Set
import time


class SitemapParser:
    """Parser pro nalezení a extrakci URL ze sitemapy"""
    
    def __init__(self, base_url: str, timeout: int = 10):
        """
        Inicializace parseru
        
        Args:
            base_url: Základní URL webu
            timeout: Timeout pro HTTP požadavky
        """
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'SEO Analyzer Bot 1.0'
        })
    
    def find_sitemap(self) -> str:
        """
        Najde sitemapu webu
        
        Zkusí:
        1. /sitemap.xml
        2. /sitemap_index.xml
        3. robots.txt (hledá Sitemap: directive)
        
        Returns:
            URL sitemapy nebo None pokud není nalezena
        """
        # Zkusit standardní cesty
        sitemap_paths = [
            '/sitemap.xml',
            '/sitemap_index.xml',
            '/sitemap1.xml'
        ]
        
        for path in sitemap_paths:
            url = urljoin(self.base_url, path)
            try:
                response = self.session.head(url, timeout=self.timeout, allow_redirects=True)
                if response.status_code == 200:
                    # Ověřit, že je to XML
                    content_type = response.headers.get('Content-Type', '').lower()
                    if 'xml' in content_type:
                        return response.url
            except requests.RequestException:
                continue
        
        # Zkusit najít v robots.txt
        robots_url = urljoin(self.base_url, '/robots.txt')
        try:
            response = self.session.get(robots_url, timeout=self.timeout)
            if response.status_code == 200:
                for line in response.text.split('\n'):
                    line = line.strip()
                    if line.lower().startswith('sitemap:'):
                        sitemap_url = line.split(':', 1)[1].strip()
                        # Ověřit, že sitemapa existuje
                        try:
                            test_response = self.session.head(sitemap_url, timeout=self.timeout)
                            if test_response.status_code == 200:
                                return sitemap_url
                        except requests.RequestException:
                            continue
        except requests.RequestException:
            pass
        
        return None
    
    def parse_sitemap(self, sitemap_url: str) -> Set[str]:
        """
        Parsuje sitemapu a extrahuje všechny URL
        
        Podporuje:
        - Obyčejné sitemapy
        - Sitemap indexy (více sitemap v jednom)
        
        Args:
            sitemap_url: URL sitemapy
            
        Returns:
            Set všech URL nalezených v sitemapě
        """
        all_urls = set()
        visited_sitemaps = set()
        
        def parse_sitemap_recursive(url: str):
            """Rekurzivní parsování sitemapy (pro sitemap indexy)"""
            if url in visited_sitemaps:
                return
            visited_sitemaps.add(url)
            
            try:
                response = self.session.get(url, timeout=self.timeout)
                response.raise_for_status()
                
                # Parsovat XML
                root = ET.fromstring(response.content)
                
                # Namespace pro sitemap XML
                namespace = {'sitemap': 'http://www.sitemaps.org/schemas/sitemap/0.9'}
                
                # Zkontrolovat, zda je to sitemap index
                sitemapindex = root.find('sitemap:sitemapindex', namespace)
                if sitemapindex is not None:
                    # Je to sitemap index - rekurzivně parsovat všechny sitemapy
                    for sitemap_elem in root.findall('sitemap:sitemap', namespace):
                        loc_elem = sitemap_elem.find('sitemap:loc', namespace)
                        if loc_elem is not None and loc_elem.text:
                            sitemap_url = loc_elem.text.strip()
                            parse_sitemap_recursive(sitemap_url)
                            time.sleep(0.5)  # Rate limiting
                else:
                    # Obyčejná sitemapa - extrahovat URL
                    for url_elem in root.findall('sitemap:url', namespace):
                        loc_elem = url_elem.find('sitemap:loc', namespace)
                        if loc_elem is not None and loc_elem.text:
                            url = loc_elem.text.strip()
                            all_urls.add(url)
                            
            except requests.RequestException as e:
                print(f"Chyba při načítání sitemapy {url}: {e}")
            except ET.ParseError as e:
                print(f"Chyba při parsování XML sitemapy {url}: {e}")
            except Exception as e:
                print(f"Neočekávaná chyba při parsování sitemapy {url}: {e}")
        
        parse_sitemap_recursive(sitemap_url)
        return all_urls
    
    def get_all_urls(self) -> List[str]:
        """
        Hlavní metoda pro získání všech URL ze sitemapy
        
        Returns:
            Seznam všech URL nalezených v sitemapě
        """
        sitemap_url = self.find_sitemap()
        if not sitemap_url:
            raise ValueError(f"Nepodařilo se najít sitemapu pro {self.base_url}")
        
        print(f"Nalezena sitemapa: {sitemap_url}")
        urls = self.parse_sitemap(sitemap_url)
        print(f"Nalezeno {len(urls)} URL v sitemapě")
        
        return sorted(list(urls))

File: netlify/functions/utils/sitemapParser.ts
import fetch from 'node-fetch';
import { parseString } from 'xml2js';
import { promisify } from 'util';

const parseXML = promisify(parseString);

export interface SitemapParserOptions {
  timeout?: number;
}

export class SitemapParser {
  private baseUrl: string;
  private timeout: number;

  constructor(baseUrl: string, options: SitemapParserOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.timeout = options.timeout || 10000;
  }

  private async fetchUrl(url: string): Promise<string> {
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'SEO Analyzer Bot 1.0'
      },
      timeout: this.timeout
    } as any);
    
    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }
    
    return await response.text();
  }

  async findSitemap(): Promise<string | null> {
    const sitemapPaths = [
      '/sitemap.xml',
      '/sitemap_index.xml',
      '/sitemap1.xml'
    ];

    // Try standard paths
    for (const path of sitemapPaths) {
      try {
        const url = `${this.baseUrl}${path}`;
        const response = await fetch(url, {
          method: 'HEAD',
          headers: {
            'User-Agent': 'SEO Analyzer Bot 1.0'
          },
          timeout: this.timeout
        } as any);
        
        if (response.ok) {
          const contentType = response.headers.get('content-type') || '';
          if (contentType.toLowerCase().includes('xml')) {
            return url;
          }
        }
      } catch (e) {
        continue;
      }
    }

    // Try robots.txt
    try {
      const robotsUrl = `${this.baseUrl}/robots.txt`;
      const robotsText = await this.fetchUrl(robotsUrl);
      
      for (const line of robotsText.split('\n')) {
        const trimmed = line.trim();
        if (trimmed.toLowerCase().startsWith('sitemap:')) {
          const sitemapUrl = trimmed.split(':', 2)[1].trim();
          // Verify it exists
          try {
            const testResponse = await fetch(sitemapUrl, {
              method: 'HEAD',
              headers: {
                'User-Agent': 'SEO Analyzer Bot 1.0'
              },
              timeout: this.timeout
            } as any);
            
            if (testResponse.ok) {
              return sitemapUrl;
            }
          } catch (e) {
            continue;
          }
        }
      }
    } catch (e) {
      // robots.txt not found or error
    }

    return null;
  }

  private async parseSitemapRecursive(
    sitemapUrl: string,
    allUrls: Set<string>,
    visitedSitemaps: Set<string>
  ): Promise<void> {
    if (visitedSitemaps.has(sitemapUrl)) {
      return;
    }
    visitedSitemaps.add(sitemapUrl);

    try {
      const xmlText = await this.fetchUrl(sitemapUrl);
      const parsed = await parseXML(xmlText) as any;

      // Check if it's a sitemap index
      if (parsed.sitemapindex) {
        const sitemaps = parsed.sitemapindex.sitemap || [];
        for (const sitemap of sitemaps) {
          if (sitemap.loc && sitemap.loc[0]) {
            const nestedUrl = sitemap.loc[0].trim();
            await this.parseSitemapRecursive(nestedUrl, allUrls, visitedSitemaps);
            await new Promise(resolve => setTimeout(resolve, 500)); // Rate limiting
          }
        }
      } else if (parsed.urlset) {
        // Regular sitemap
        const urls = parsed.urlset.url || [];
        for (const url of urls) {
          if (url.loc && url.loc[0]) {
            allUrls.add(url.loc[0].trim());
          }
        }
      }
    } catch (error) {
      console.error(`Error parsing sitemap ${sitemapUrl}:`, error);
    }
  }

  async getAllUrls(): Promise<string[]> {
    const sitemapUrl = await this.findSitemap();
    if (!sitemapUrl) {
      throw new Error(`Nepodařilo se najít sitemapu pro ${this.baseUrl}`);
    }

    const allUrls = new Set<string>();
    const visitedSitemaps = new Set<string>();

    await this.parseSitemapRecursive(sitemapUrl, allUrls, visitedSitemaps);

    return Array.from(allUrls).sort();
  }
}

